clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2162,8 +2171,10 @@ public void close(Duration timeout) {
         acquire();
         try {
             if (!closed) {
-                closed = true;
+                // need to close before setting the flag since the close function
+                // itself may trigger rebalance callback that needs the consumer to be open still
                 close(timeout.toMillis(), false);
+                closed = true;
```

Comment:
Should we move this to the `finally` block? If there is an exception raised from `close`, do we want to allow retry?

Reply:
Good point.